File: 0402-remove-k-digits/0402-remove-k-digits.java
class Solution {
    public String removeKdigits(String num, int k) {
        Stack<Character> stack = new Stack<>();
        int n = num.length();
        String result = "";
        for(int i=0;i<n;i++){
            
            char ele = num.charAt(i);
            while(!stack.isEmpty() && k> 0 && ((int)stack.peek()) > (int) ele ){
                stack.pop();
                k-=1;
            }
            stack.push(ele);
            
        }
        while(!stack.isEmpty() && k> 0 ){
                stack.pop();
                k-=1;
            }
        while(!stack.isEmpty()){
            result = result + stack.pop();
        }
        if(result.length() == 0)
            return "0";
        


        result =  new StringBuilder(result).reverse().toString();
        int st = 0;
        while(st < result.length() && result.charAt(st) == '0') st++;
        return st < result.length() ? result.substring(st): "0";
    }
}

File: 0402-remove-k-digits/0402-remove-k-digits.py
class Solution:
    def removeKdigits(self, num: str, k: int) -> str:
        stack = []
        result = ""
        for i in num :
            while k and stack and ord(stack[-1]) - ord("0") > ord(i) - ord("0") :
                k-=1
                stack.pop()
            if not stack and i == "0":
                continue
            stack.append(i)
        while stack and k :
            stack.pop()
            k-=1
        for i in stack :
            result+=i
        if not result:
            return "0"
        return result
        
        